src/metabase/models/card.clj

```diff
@@ -601,7 +665,13 @@ saved later when it is ready."
                                :parameters :parameter_mappings :collection_id :collection_position :cache_ttl]
          card-data            (assoc (zipmap data-keys (map card-data data-keys))
                                      :creator_id (:id creator)
-                                     :dataset (boolean (:dataset card-data))
+                                     :dataset    (or (and (:type card-data) (is-model? card-data))
+                                                     (boolean (:dataset card-data)))
+                                     :type       (or (:type card-data)
+                                                     ;; most tests will create a model using :dataset true, so we'll
+                                                     ;; respect that
+                                                     (when (:dataset card-data) "model")
+                                                     "question")
```

Comment:
Why don't we reuse `ensure-type-and-dataset-are-consistent` here?

Reply:
do you mean to set these values or as a check?

`ensure-type-and-dataset-are-consistent` will eventually be called in the `before-insert` hook so it's a bit redundant.

Reply:
To set the values. From what I can tell this code duplicates logic from there.

It's not redundant because it looks like there's currently a bug on line 679 (I can't highlight it in github) where `dataset` is used but it should be the `dataset` that comes from `ensure-type-and-dataset-are-consistent`.

```
         result-metadata-chan (result-metadata-async {:query    dataset_query
                                                      :metadata result_metadata
                                                      :dataset? dataset})
```

Also where are the tests for this? I only see API and revision tests, no tests on the create-card! function itself.

Reply:
right, I updated how we set the values here: 3f5fd80bb0


On testing, I don't think we have anything to test regarding the `dataset?` flag there since it's used to make sure we preserve the original metadata when updating the card. Since this is during card creation, there is no "original metadata"

Reply:
Ah cool thanks for explaining.

Continuing the conversation [here](https://github.com/metabase/metabase/pull/38074/files#r1475028148) because the referenced code is out of date